Plutarch/Repr/Data.hs

```diff
@@ -136,7 +163,32 @@ instance
   pmatch' x f =
     pmatch x (f . DeriveAsDataRec . SOP.to . SOP.hcoerce . SOP . (Z @_ @_ @'[]) . unPRec . unPDataRec)
 
--- | @since 1.10.0
+{- |
+@DeriveAsDataStruct@ derives @PlutusType@ instances for the given type as Data structure, namely, using @Constr@ constructor
+of the @Data@ type. Each constructor of the given type will have matching constructor index in the order of its definition.
+
+Also, it is important to note that each fields can only contain term that has innermost representation of Data. Hence,
+@PInteger@ is not allowed but @PAsData PInteger@ is allowed. Failure to follow this requirement will result in type error
+with detailed explanation of the issue.
+
+@PInner@ of defined type will be @PDataStruct (struct :: [[S -> Type]])@ where @struct@ is SOP type of its structure. Since @PInner@ of
+@PDataStruct@ is @PData@, multiple data encoded structure can be nested without being wrapped in @PAsData@.
+
+Consult example below for defining custom data-encoded datatype:
+@@
+data PBobData (a :: S -> Type) (s :: S)
+  = PBobData (Term s (PAsData a)) (Term s (PAsData PBool))
+  | PRobData (Term s (PAsData PByteString))
+  deriving stock (Generic)
+  deriving anyclass (SOP.Generic)
+  deriving PlutusType via (DeriveAsDataStruct (PBobData a))
+
+pcon $ PBobData (pdata 10) (pdata pfalse) -- Constr 0 [#10, #false]
+pcon $ PBobData "hello" -- Constr 1 [#"hello"]
```

Comment:
```suggestion
pcon $ PBobData (pdata 10) (pdata pfalse) -- Constr 0 [#10, #false]
pcon $ PRobData "hello" -- Constr 1 [#"hello"]
```

would be cool if there was some doctest tool usable for Haskell but unfortunately

Reply:
Yeah... I saw something similar but it was pretty janky to set up